streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java

```diff
@@ -26,9 +26,9 @@
 import static org.apache.kafka.streams.kstream.internals.WrappingNullableUtils.prepareKeyDeserializer;
 import static org.apache.kafka.streams.kstream.internals.WrappingNullableUtils.prepareValueDeserializer;
 
-public class SourceNode<KIn, VIn, KOut, VOut> extends ProcessorNode<KIn, VIn, KOut, VOut> {
+public class SourceNode<KIn, VIn> extends ProcessorNode<KIn, VIn, KIn, VIn> {
```

Comment:
Here, we declare that the source node can only forward the same type it receives.